Sources/Controllers/MessagesViewController+Keyboard.swift

```diff
@@ -81,23 +92,32 @@ extension MessagesViewController {
         }
     }
 
-    // MARK: - Helpers
+    private func computeScrollViewBottomInset(forKeyboardFrame keyboardFrame: CGRect) -> CGFloat {
+        // we only need to adjust for the part of the keyboard that covers (i.e. intersects) our collection view
+        let intersection = messagesCollectionView.frame.intersection(keyboardFrame)
+        
+        if intersection.isNull || intersection.maxY < messagesCollectionView.frame.maxY - 1e-6 /* never compare floats without a tolerance */ {
```

Comment:
What's the purpose here to compare `intersection.maxY < messagesCollectionView.frame.maxY - 1e-6`?
Is this situation, `intersection.maxY == min(messagesCollectionView.frame.maxY, keyboardFrame.maxY`, I think we only need to check `isNull`, if it fails, go to `else` statement.

Reply:
The equality does not always hold. The undocked keyboard still generates frame change notifications, and they go through the same code, so it's important to only apply the inset when the keyboard covers the bottom of the collection view.

I do agree that the check feels a bit weird. I've tried a different approach, one recommended by Apple: track whether the keyboard is visible using show/hide events, and ignore frame changes when it's not visible. (The undocked keyboard is considered hidden.) Unfortunately, I couldn't get this to work properly, because the input bar is a keyboard accessory, so it's crucial to differentiate between 2 cases — the keyboard being hidden (and the input bar being at the bottom) and the keyboard being undocked (and the input bar being tied to the keyboard). It turns out that just using the keyboard frame in all cases allows us to handle them all transparently without needing to branch.

I'll add a comment about this to the code.

Reply:
@andreyvit  So what's problem if we remove `intersection.maxY < messagesCollectionView.frame.maxY - 1e-6` check?

Reply:
Re-raising this comment so it's not hidden